napalm_logs/config/junos/BGP_NEIGHBOR_STATE_CHANGED.yml

```diff
@@ -0,0 +1,209 @@
+# This error tag corresponds to syslog messages notifying that the configured
+# bgp neighbor has changed state
+messages:
+  # 'error' should be unique and vendor agnostic. Currently we are using the JUNOS syslog message name as the canonical name.
+  # This may change if we are able to find a more well defined naming system.
+  - error: BGP_NEIGHBOR_STATE_CHANGED
+    tag: RPD_BGP_NEIGHBOR_STATE_CHANGED
+    values:
+      peer: ([\w\d:\.]+)
+      asn: (\d+)
+      peeringType: (\w+)
+      event: (\w+)
+    line: 'BGP peer {peer} ({peeringType} AS {asn}) changed state from Idle to Connect (event {event}) (instance master)'
+    model: openconfig-bgp
+    mapping:
+      variables:
+        bgp//neighbors//neighbor//{peer}//state//peer_as: asn
+        bgp//neighbors//neighbor//{peer}//state//session-state-change-event: event
+      static:
+        bgp//neighbors//neighbor//{peer}//state//session-state: CONNECT
+        bgp//neighbors//neighbor//{peer}//state//session-state-old: IDLE
+
+  - error: BGP_NEIGHBOR_STATE_CHANGED
+    tag: RPD_BGP_NEIGHBOR_STATE_CHANGED
+    values:
+      peer: ([\w\d:\.]+)
+      asn: (\d+)
+      peeringType: (\w+)
+      event: (\w+)
+    line: 'BGP peer {peer} ({peeringType} AS {asn}) changed state from Connect to Idle (event {event}) (instance master)'
+    model: openconfig-bgp
+    mapping:
+      variables:
+        bgp//neighbors//neighbor//{peer}//state//peer_as: asn
+        bgp//neighbors//neighbor//{peer}//state//session-state-change-event: event
+      static:
+        bgp//neighbors//neighbor//{peer}//state//session-state: IDLE
```

Comment:
The repeat entries can be condensed into one entry by matching the old and new states and then converting the values into the appropriate openconfig names. You can do this by casting to a function. See https://github.com/napalm-automation/napalm-logs/blob/master/napalm_logs/config/junos/SYSTEM_ALARM.yml#L13 for an example.

In this case you could add a function to convert any changes like OpenSent to OPEN_SENT and others you could just return as uppercase, i.e:

```
def bgp_state_convert(state):
    state_dict = {'OpenSent': 'OPEN_SENT',
                         'OpenConfirm': 'OPEN_CONFIRM'}
    return state_dict.get(state, state.upper())
```

Reply:
I think this is better than my current approach but IMO this is going to get unwieldy fast once you start to throw other vendors. For example, Juniper calls it `OpenSent` while cisco calls it `OPEN SENT`. Obviously these mappings have to be defined somewhere but I don't think hiding them in a function inside the utils package is the best idea in the long run. 

Like I said, this is better than what I have for now, but I think it is worth further discussion, along with the vendor event message mappings.

Reply:
We should probably have a functionality directly into the YAML profiles and help with this kind of 1:1 mappings, e.g.,

```yaml
mapping:
  replace:
    state:
      OpenSent: OPEN_SENT
      OpenConfirm: OPEN_CONFIRM
  variables:
    bgp//neighbors//neighbor//{peer}//state//peer_as: asn
    bgp//neighbors//neighbor//{peer}//state//session-state-change-event: event
    bgp//neighbors//neighbor//{peer}//state//session-state: state
```

Any thoughts on this @loverend @lampwins?

Anyway, for the moment I would say to get this PR in, and investigate this sort of optimisations later (before they'll grow more), but it's very good to have this discussion right now.

Reply:
Or maybe even something like:

```yaml
preprocess:
  state:
    - replace:
        OpenSent: OPEN_SENT
        OpenConfirm: OPEN_CONFIRM
    - upper
    - lower
    - upper
mapping:
  variables:
    bgp//neighbors//neighbor//{peer}//state//peer_as: asn
    bgp//neighbors//neighbor//{peer}//state//session-state-change-event: event
    bgp//neighbors//neighbor//{peer}//state//session-state: state
```

Just in case we'd even need more extensive preprocessing to some fields (that's maybe just not the best example, but I think you see what I mean: to a certain field you can apply several consecutive operations).